ddtrace/sampler.py

```diff
@@ -92,3 +103,251 @@ def set_sample_rate_by_service(self, rate_by_service):
 
 # Default key for service with no specific rate
 RateByServiceSampler._default_key = RateByServiceSampler._key()
+
+
+class DatadogSampler(BaseSampler):
+    """
+    """
+    __slots__ = ('rules', 'rate_limit')
+
+    DEFAULT_RATE_LIMIT = 100
+    NO_RATE_LIMIT = -1
+
+    def __init__(self, rules=None, rate_limit=None):
+        """
+        Constructor for DatadogSampler sampler
+
+        :param rules: List of :class:`SamplingRule` rules to apply to the root span of every trace, default no rules
+        :type rules: :obj:`list` of :class:`SamplingRule`
+        :param rate_limit: Global rate limit (traces per second) to apply to all traces regardless of the rules
+            applied to them, default 100 traces per second
+        :type rate_limit: :obj:`int`
+        """
+        # Ensure rules is a list
+        if not rules:
+            rules = []
+
+        # Validate that the rules is a list of SampleRules
+        for rule in rules:
+            if not isinstance(rule, SamplingRule):
+                raise TypeError('Rule {!r} must be a sub-class of type ddtrace.sampler.SamplingRules'.format(rule))
+        self.rules = rules
+
+        # Configure rate limiter
+        if rate_limit is None:
+            rate_limit = self.DEFAULT_RATE_LIMIT
+        self.limiter = RateLimiter(rate_limit)
+
+    def sample(self, span):
+        """
+        Decide whether the provided span should be sampled or not
+
+        The span provided should be the root span in the trace.
+
+        :param span: The root span of a trace
+        :type span: :class:`ddtrace.span.Span`
+        :returns: Whether the span was sampled or not
+        :rtype: :obj:`bool`
+        """
+        # If there are rules defined, then iterate through them and find one that wants to sample
+        matching_rule = None
+        if self.rules:
+            # Go through all rules and grab the first one that matched
```

Comment:
I would prefer just going through all rules even when there is a match since it would enable merging matched rules later on when we expand this API. That is unless we don't want that. Basically I think the order should be determined by whether we want to allow multiple rules to apply in a later iteration.

Reply:
I think this way is much easier to communicate to customers, and unless we know exact use cases for why we need multiple rules to match, then we should not think about that just yet.

> The first rule that matches is applied. So it is suggested to order them from most specific to least specific.

Reply:
The point was mostly that by going with the first match we lose the flexibility, while not gaining anything. I don't see a first match as easier to understand, but it's definitely more performant.

Reply:
I'm not sure about the purpose of `merging matched rules`. What do you actually want to merge? 
Plus, technically, shouldn't we just care about the most granular rule?